### PR TITLE
CSPL-1412/CSPL-1413 Break down app framework tests

### PR DESCRIPTION
**Pull request by jambrosiano**
Breakdown of App Framework test for C3 and M4 SVAs
This PR break down existing App framework tests to reduce flakiness.
New tests will perform only 1 action (for ex Upgrade, downgrade apps, or scaling pods) then will verify and finish, whereas before 1 test would perform both upgrade and downgrade of the apps.

Test runs:
`[2] • [SLOW TEST:1704.854 seconds]
[2] c3appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[2]   Single Site Indexer Cluster with SHC (C3) with App Framework
[2]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:63
[2]     integration, c3, appframework: can deploy a C3 SVA with App Framework enabled, install apps and upgrade them
[2]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:64
[2] ------------------------------
[2] {"level":"info","ts":1635462825.937587,"msg":"testenv deleted.\n","testenv":"c3appfw-jhu"}
[2] {"level":"info","ts":1635462825.93763,"msg":"testenv deleted.\n","testenv":"c3appfw-jhu"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-jhu_junit.xml
[2]
[2] Ran 1 of 1 Specs in 1721.059 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS

[1] • [SLOW TEST:1741.730 seconds]
[1] c3appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[1]   Single Site Indexer Cluster with SHC (C3) with App Framework
[1]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:258
[1]     smoke, c3, appframework: can deploy a C3 SVA with App Framework enabled, install apps and downgrade them
[1]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:259
[1] ------------------------------
[1] {"level":"info","ts":1635465067.0392609,"msg":"testenv deleted.\n","testenv":"c3appfw-kcm"}
[1] {"level":"info","ts":1635465067.039286,"msg":"testenv deleted.\n","testenv":"c3appfw-kcm"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-kcm_junit.xml
[1]
[1] Ran 1 of 2 Specs in 1757.840 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 1 Pending | 0 Skipped
[1] PASS

[2] • [SLOW TEST:755.196 seconds]
[2] c3appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[2]   Clustered deployment (C3 - clustered indexer, search head cluster)
[2]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:736
[2]     smoke, c3, appframework: can deploy a C3 SVA and have apps installed locally on CM and SHC Deployer
[2]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:737
[2] ------------------------------
[2] {"level":"info","ts":1635470257.10548,"msg":"testenv deleted.\n","testenv":"c3appfw-g05"}
[2] {"level":"info","ts":1635470257.105502,"msg":"testenv deleted.\n","testenv":"c3appfw-g05"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-g05_junit.xml
[2]
[2] Ran 1 of 4 Specs in 771.501 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 3 Pending | 0 Skipped
[2] PASS

[1] • [SLOW TEST:1563.236 seconds]
[1] c3appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[1]   Clustered deployment (C3 - clustered indexer, search head cluster)
[1]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:854
[1]     integration, c3, appframework: can deploy a C3 SVA and have ES app installed on SHC
[1]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:855
[1] ------------------------------
[1] {"level":"info","ts":1635472266.5344708,"msg":"testenv deleted.\n","testenv":"c3appfw-v4o"}
[1] {"level":"info","ts":1635472266.534501,"msg":"testenv deleted.\n","testenv":"c3appfw-v4o"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-v4o_junit.xml
[1]
[1] Ran 1 of 5 Specs in 1580.172 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 4 Pending | 0 Skipped
[1] PASS

[1] • [SLOW TEST:1003.444 seconds]
[1] c3appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[1]   Clustered deployment (C3 - clustered indexer, search head cluster)
[1]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:968
[1]     c3, integration, appframework: can deploy a C3 SVA with apps installed locally on CM and Deployer, cluster-wide on Peers and SHs, then upgrade them
[1]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:969
[1] ------------------------------
[1] {"level":"info","ts":1635533319.94327,"msg":"testenv deleted.\n","testenv":"c3appfw-e29"}
[1] {"level":"info","ts":1635533319.943298,"msg":"testenv deleted.\n","testenv":"c3appfw-e29"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-e29_junit.xml
[1]
[1] Ran 1 of 6 Specs in 1019.295 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 5 Pending | 0 Skipped
[1] PASS

[2] • [SLOW TEST:1295.427 seconds]
[2] c3appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:29
[2]   Clustered deployment (C3 - clustered indexer, search head cluster)
[2]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:1157
[2]     c3, integration, appframework: can deploy a C3 SVA with apps installed locally on CM and Deployer, cluster-wide on Peers and SHs, then downgrade them
[2]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/appframework_test.go:1158
[2] ------------------------------
[2] {"level":"info","ts":1635532174.483169,"msg":"testenv deleted.\n","testenv":"c3appfw-dot"}
[2] {"level":"info","ts":1635532174.483247,"msg":"testenv deleted.\n","testenv":"c3appfw-dot"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/c3/appframework/c3appfw-dot_junit.xml
[2]
[2] Ran 1 of 7 Specs in 1311.181 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 6 Pending | 0 Skipped
[2] PASS
`
For M4, break down was already performed so this PR mainly adds test steps for clarity (further improving of logging messages to be performed in another PR soon)

Adding Break down of test for S1 SVA to this PR (CSPL-1413)
Previously 1 test, now cut in 3 parts: 1 upgrade, 1 downgrade, 1 scaling
test runs:
`[2] • [SLOW TEST:900.288 seconds]
[2] s1appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:30
[2]   Standalone deployment (S1) with App Framework
[2]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:64
[2]     smoke, s1, appframework: can deploy a standalone instance with App Framework enabled, install apps and upgrade them
[2]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:65
[2] ------------------------------
[2] {"level":"info","ts":1635791988.216133,"msg":"testenv deleted.\n","testenv":"s1appfw-vfm"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/s1appfw-vfm_junit.xml
[2]
[2] Ran 1 of 1 Specs in 915.824 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS

[1] • [SLOW TEST:969.847 seconds]
[1] s1appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:30
[1]   Standalone deployment (S1) with App Framework
[1]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:222
[1]     smoke, s1, appframework: can deploy a standalone instance with App Framework enabled, install apps and downgrade them
[1]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:223
[1] ------------------------------
[1] {"level":"info","ts":1635794737.356558,"msg":"testenv deleted.\n","testenv":"s1appfw-cdw"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/s1appfw-cdw_junit.xml
[1]
[1] Ran 1 of 2 Specs in 985.836 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 1 Pending | 0 Skipped
[1] PASS

[1] • [SLOW TEST:1028.380 seconds]
[1] s1appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:30
[1]   Standalone deployment (S1) with App Framework
[1]   /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:387
[1]     smoke, s1, appframework: can deploy a Standalone instance with App Framework enabled, install apps, scale up, install apps on new pod, scale down
[1]     /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/appframework_test.go:388
[1] ------------------------------
[1] {"level":"info","ts":1635799145.904993,"msg":"testenv deleted.\n","testenv":"s1appfw-plr"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1412/splunk-operator/test/s1/appframework/s1appfw-plr_junit.xml
[1]
[1] Ran 1 of 3 Specs in 1044.317 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 2 Pending | 0 Skipped
[1] PASS
`

### CODE REVIEW

**Review by tpereirasplunk — APPROVED** [review]

**Review by pdhanoyasplunk — APPROVED** [review]

**Review by smohan-splunk — APPROVED** [review]

File: test/c3/appframework/appframework_test.go [hunk 1]
```diff
@@ -407,28 +378,228 @@ var _ = Describe("c3appfw test", func() {
...
+			// Deploy MC CRD
+			testenvInstance.Log.Info("Deploy Monitoring Console")
+			mcName := deployment.GetName()
+			mc, err := deployment.DeployMonitoringConsoleWithGivenSpec(testenvInstance.GetName(), mcName, mcSpec)
+			Expect(err).To(Succeed(), "Unable to deploy Monitoring Console instance")
+
+			// Verify MC is Ready and stays in ready state
+			testenv.VerifyMonitoringConsoleReady(deployment, deployment.GetName(), mc, testenvInstance)
+
+			// Create App framework Spec for C3
```
Comment (kashok-splunk) [inline test/c3/appframework/appframework_test.go:522]: Missing upload of V1 apps for C3 deployment @pdhanoyasplunk @jambrosiano

File: test/s1/appframework/appframework_test.go [hunk 2]
```diff
@@ -53,61 +53,84 @@ var _ = Describe("s1appfw test", func() {
...
+			testenvInstance.Log.Info("Deploy Monitoring Console")
 			mcName := deployment.GetName()
+			mc, err := deployment.DeployMonitoringConsoleWithGivenSpec(testenvInstance.GetName(), mcName, mcSpec)
+			Expect(err).To(Succeed(), "Unable to deploy Monitoring Console One instance")
 
 			// Upload V1 apps to S3
 			s3TestDir = "s1appfw-" + testenv.RandomDNSName(4)
-			appFileList := testenv.GetAppFileList(appListV1, 1)
-			uploadedFiles, err := testenv.UploadFilesToS3(testS3Bucket, s3TestDir, appFileList, downloadDirV1)
+			appFileList = testenv.GetAppFileList(appListV1, 1)
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:114]: Redundant line @jambrosiano @pdhanoyasplunk

File: test/s1/appframework/appframework_test.go [hunk 3]
```diff
@@ -121,43 +144,131 @@ var _ = Describe("s1appfw test", func() {
...
 			// Wait for Standalone to be in READY status
 			testenv.StandaloneReady(deployment, deployment.GetName(), standalone, testenvInstance)
 
+			// Verify apps are downloaded by init-container on Standalone Pod and MC
+			appVersion := "V1"
+			initContDownloadLocationStandalonePod := "/init-apps/" + appSourceName
+			initContDownloadLocationMCPod := "/init-apps/" + appSourceNameMC
+			standalonePodName := fmt.Sprintf(testenv.StandalonePod, deployment.GetName(), 0)
+			mcPodName := fmt.Sprintf(testenv.MonitoringConsolePod, mcName, 0)
+			appFileList = testenv.GetAppFileList(appListV1, 1)
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:160]: Redundant line @jambrosiano @pdhanoyasplunk

File: test/s1/appframework/appframework_test.go [hunk 4]
```diff
@@ -121,43 +144,131 @@ var _ = Describe("s1appfw test", func() {
...
+			podNames := []string{standalonePodName, mcPodName}
+			testenvInstance.Log.Info("Verify V1 apps are downloaded by init container for Standalone", "POD", standalonePodName, "version", appVersion)
+			testenv.VerifyAppsDownloadedByInitContainer(deployment, testenvInstance, testenvInstance.GetName(), []string{standalonePodName}, appFileList, initContDownloadLocationStandalonePod)
+			testenvInstance.Log.Info("Verify V1 apps are downloaded by init container for Monitoring Console", "POD", mcPodName, "version", appVersion)
+			testenv.VerifyAppsDownloadedByInitContainer(deployment, testenvInstance, testenvInstance.GetName(), []string{mcPodName}, appFileList, initContDownloadLocationMCPod)
+
+			// Verify V1 apps are copied to location
+			testenvInstance.Log.Info("Verify V1 apps are copied to correct location on Pod for app", "version", appVersion)
+			testenv.VerifyAppsCopied(deployment, testenvInstance, testenvInstance.GetName(), podNames, appListV1, true, true)
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:169]: `clusterWideInstall` should be set to `false`

File: test/s1/appframework/appframework_test.go [hunk 5]
```diff
@@ -121,43 +144,131 @@ var _ = Describe("s1appfw test", func() {
...
+
+			// Upload newer version of apps (V2) to S3 for Standalone and Monitoring Console
+			appFileList = testenv.GetAppFileList(appListV2, 2)
+			appVersion = "V2"
+			uploadedFiles, err = testenv.UploadFilesToS3(testS3Bucket, s3TestDir, appFileList, downloadDirV2)
+			Expect(err).To(Succeed(), "Unable to upload apps to S3 test directory")
+			uploadedApps = append(uploadedApps, uploadedFiles...)
+			uploadedFiles, err = testenv.UploadFilesToS3(testS3Bucket, s3TestDirMC, appFileList, downloadDirV2)
+
 			// Upload Apps to S3 for MC
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:188]: Remove line 188-194, redundant code

File: test/s1/appframework/appframework_test.go [hunk 6]
```diff
@@ -121,43 +144,131 @@ var _ = Describe("s1appfw test", func() {
...
+			s3TestDirMC = "s1appfw-mc-" + testenv.RandomDNSName(4)
 			appFileList = testenv.GetAppFileList(appListV1, 1)
 			uploadedFiles, err = testenv.UploadFilesToS3(testS3Bucket, s3TestDirMC, appFileList, downloadDirV1)
 			Expect(err).To(Succeed(), "Unable to upload apps to S3 test directory")
 			uploadedApps = append(uploadedApps, uploadedFiles...)
 
+			// Wait for the poll period for the apps to be downloaded
+			time.Sleep(2 * time.Minute)
+
+			// Wait for Standalone to be in UPDATING status
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:198]: Check comment "PENDING" status

File: test/s1/appframework/appframework_test.go [hunk 7]
```diff
@@ -121,43 +144,131 @@ var _ = Describe("s1appfw test", func() {
...
+
+			// Verify V2 apps are downloaded by init-container
+			testenvInstance.Log.Info("Verify V2 apps are downloaded by init container for apps", "POD", standalonePodName, "version", appVersion)
+			testenv.VerifyAppsDownloadedByInitContainer(deployment, testenvInstance, testenvInstance.GetName(), []string{standalonePodName}, appFileList, initContDownloadLocationStandalonePod)
+			testenvInstance.Log.Info("Verify V2 apps are downloaded by init container for apps", "POD", mcPodName, "version", appVersion)
+			testenv.VerifyAppsDownloadedByInitContainer(deployment, testenvInstance, testenvInstance.GetName(), []string{mcPodName}, appFileList, initContDownloadLocationMCPod)
+
+			// Verify V2 apps are copied to location
+			testenvInstance.Log.Info("Verify V2 apps are copied to correct location on Pod for app", "version", appVersion)
+			testenv.VerifyAppsCopied(deployment, testenvInstance, testenvInstance.GetName(), podNames, appListV2, true, true)
```
Comment (kashok-splunk) [inline test/s1/appframework/appframework_test.go:215]: Change check to false `testenv.VerifyAppsCopied(deployment, testenvInstance, testenvInstance.GetName(), podNames, appListV2, true, false)`

**Pull request merged**